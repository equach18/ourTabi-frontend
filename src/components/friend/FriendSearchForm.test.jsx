import { screen, fireEvent } from "@testing-library/react";
import FriendSearchForm from "./FriendSearchForm";
import OurTabiApi from "../../api/ourTabiApi";
import {
  renderWithUserProvider,
  testUser1,
  testUser2,
  testUser3,
  testFriendsData,
} from "../../test-helpers/_testCommons";

vi.mock("../../api/ourTabiApi");

const testUserContext = {
  currentUser: testUser1,
  ...testFriendsData,
};

describe("FriendSearchForm", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("matches snapshot for default empty state", () => {
    const { asFragment } = renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot while searching", () => {
    const pendingPromise = new Promise(() => {});
    OurTabiApi.searchUsers.mockReturnValueOnce(pendingPromise);

    const { asFragment } = renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "waiting" },
    });
    fireEvent.click(screen.getByRole("button", { name: /search/i }));

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with input and button", () => {
    renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });

    expect(
      screen.getByPlaceholderText(/search by username/i)
    ).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /search/i })).toBeInTheDocument();
  });

  it("displays an error message", async () => {
    OurTabiApi.searchUsers.mockRejectedValue(new Error("An error!!"));

    renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "fail" },
    });
    fireEvent.click(screen.getByRole("button", { name: /search/i }));

    expect(
      await screen.findByText(/something went wrong/i)
    ).toBeInTheDocument();
  });

  it("displays matching users", async () => {
    OurTabiApi.searchUsers.mockResolvedValue([testUser2, testUser3]);

    renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "test" },
    });
    fireEvent.click(screen.getByRole("button", { name: /search/i }));

    expect(await screen.findByText("@testUser2")).toBeInTheDocument();
    expect(await screen.findByText("@testUser3")).toBeInTheDocument();
  });

  it("displays message when no users are found", async () => {
    OurTabiApi.searchUsers.mockResolvedValue([]);

    renderWithUserProvider(<FriendSearchForm />, {
      userContext: testUserContext,
    });

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "no match" },
    });
    fireEvent.click(screen.getByRole("button", { name: /search/i }));

    expect(await screen.findByText(/no new users found/i)).toBeInTheDocument();
  });
});
