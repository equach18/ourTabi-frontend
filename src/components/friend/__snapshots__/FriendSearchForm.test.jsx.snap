// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FriendSearchForm > matches snapshot for default empty state 1`] = `
<DocumentFragment>
  <div
    class="mt-4"
  >
    <form
      class="flex items-center gap-2 mb-4"
    >
      <input
        class="flex-1 p-2 border rounded"
        placeholder="Search by username"
        required=""
        type="text"
        value=""
      />
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`FriendSearchForm > matches snapshot while searching 1`] = `
<DocumentFragment>
  <div
    class="mt-4"
  >
    <form
      class="flex items-center gap-2 mb-4"
    >
      <input
        class="flex-1 p-2 border rounded"
        placeholder="Search by username"
        required=""
        type="text"
        value="waiting"
      />
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
        type="submit"
      >
        Search
      </button>
    </form>
    <div
      class="flex justify-center items-center"
    >
      <div
        class="w-6 h-6 border-4 border-blue-500 border-t-transparent rounded-full animate-spin"
      />
    </div>
  </div>
</DocumentFragment>
`;
